### PR TITLE
[tasks] add the display of configured tasks when executing 'configure tasks...'

### LINKED ISSUE #5468: [tasks] update 'configure tasks' command and menu
**Description**

The command and menu item `Tasks: Configure Tasks...` does not display 'configured' tasks. Despite there being a **tasks.json** in the workspace, when triggering the command, no tasks are found. If however I attempt to run `Run Task...`, I can see a list of tasks with the **configure** icon.

**Steps To Reproduce**

1. open the workspace **test-resources** from `theia/packages/task/test-resources`
2. trigger the command `Tasks: Configure Tasks...` using the quick-command menu ( <kbd>F1</kbd> )
3. notice no tasks are found
4. trigger the command `Tasks: Run Tasks...` using the quick-command menu ( <kbd>F1</kbd> )
5. notice tasks are listed, with the **configure** icon
6. click the **configure** icon, the `tasks.json` is opened

### PR DESCRIPTION
**Pull request by vince-fugnitto**
#### What it does

Fixed #5468

- added _configured_ tasks when executing the command and menu item for _configure tasks..._
- previously, only _provided_ tasks where being displayed which is inconsistent with VSCode
- better multi-root support for _configured_ tasks (display the root as a description in the quick-open)
- ensure that _provided_ and _configured_ tasks are unique (no duplicates found, configured and duplicate tasks should not overlap)

#### How to test

- use the `npm` vscode-builtin-extension plugin when testing
- open a workspace (ex: `theia/packages/console)
- execute the command _Run Task..._, notice the correct source `npm` for the tasks
- execute the command _Configure Task_, notice the same correct source `npm` for tasks
- with the _Configure Task_ quick-open menu present, configure a task (this should open the `tasks.json` file with a new entry
- execute the command _Run Task..._ and verify the newly configured task is found under 'configured' tasks
- excute the command _Configure Task_, only one entry for the previously configured task should be present
- repeat the steps for a multi-root workspace

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

Signed-off-by: Vincent Fugnitto <vincent.fugnitto@ericsson.com>

### CODE REVIEW

**Comment by vince-fugnitto** [comment]
Just force-pushed to re-trigger the Gitpod build :)

**Comment by elaihau** [comment]
Did a quick test.
- it worked properly when I opened a folder as my workspace, however
- it didn't not work in the multi root setting.

**Comment by vince-fugnitto** [comment]
> Did a quick test.
> 
> * it worked properly when I opened a folder as my workspace, however
> * it didn't not work in the multi root setting.

I'll try on master, perhaps it never worked in a multi-root use case since it doesn't know where to look for the `tasks.json` :(

File: packages/task/src/browser/quick-open-task.ts [hunk 1]
```diff
@@ -160,15 +160,16 @@ export class QuickOpenTask implements QuickOpenModel, QuickOpenHandler {
...
-        if (!providedTasks.length) {
+        if (!configuredTasks && !providedTasks.length) {
             this.items.push(new QuickOpenItem({
                 label: 'No tasks found',
                 run: (_mode: QuickOpenMode): boolean => false
             }));
         }
 
-        providedTasks.forEach(task => {
+        [...configuredTasks, ...providedTasks].forEach(task => {
```
Comment (RomanNikitenko) [inline packages/task/src/browser/quick-open-task.ts]: I think about use case when at first user has some detected task, then user configure the task.
Will 'Configure Tasks' action display both tasks after that? 
So two tasks with the same label will be displayed.
Is it expected behavior?

**Comment by RomanNikitenko** [comment]
> > Did a quick test.
> > 
> > * it worked properly when I opened a folder as my workspace, however
> > * it didn't not work in the multi root setting.
> 
> I'll try on master, perhaps it never worked in a multi-root use case since it doesn't know where to look for the `tasks.json` :(

Unfortunately you are right https://github.com/theia-ide/theia/issues/4919

File: packages/task/src/browser/quick-open-task.ts (hunk 1 above)
Comment (vince-fugnitto) [reply packages/task/src/browser/quick-open-task.ts]: @RomanNikitenko what do you believe is the correct behavior?
The `configure tasks...` menu should display all configured tasks, and provided (detected tasks) which have yet to be configured?

This means if the provided task is configured, it'll no longer show in the menu as detected?

Comment (RomanNikitenko) [reply packages/task/src/browser/quick-open-task.ts]: vscode doesn't display both configurations for this case.
"Run Task" menu of Theia displays only configured configuration if both are present (configured and detected)

Comment (vince-fugnitto) [reply packages/task/src/browser/quick-open-task.ts]: Thanks! I'll wait for your PR #5313 to resolve before continuing.

**Comment by elaihau** [comment]
this one should be unblocked once https://github.com/theia-ide/theia/pull/5777 is merged.

**Comment by vince-fugnitto** [comment]
> this one should be unblocked once #5777 is merged.

Sounds good @elaihau, once #5777 is merged I'll revisit this PR again to re-confirm if it works and update it if required :)

**Comment by vince-fugnitto** [comment]
@elaihau I've rebased the PR to take advantage of the changes you made :)
Do you mind re-testing for me?

**Comment by RomanNikitenko** [comment]
I tested the changes after updating the PR.
Looks like `type`, not `source` field is used at displaying of configurations 

actual_tasks

Before the changes:
configure_before

**Comment by vince-fugnitto** [comment]
> Looks like `type`, not `source` field is used at displaying of configurations

@RomanNikitenko thank you for reviewing, I now perform the same operation to determine the `label` as the `TaskRunQuickOpenItem`.

**Comment by RomanNikitenko** [comment]
@vince-fugnitto 
I see that you updated the PR, but for me the items are displayed in the same way as before updating. Please see the video: https://youtu.be/FodCrEOrtw0

Is it something wrong with my assembly?
Could you check it?
Thank you!

**Comment by vince-fugnitto** [comment]
> I see that you updated the PR, but for me the items are displayed in the same way as before updating.

@RomanNikitenko thank you for pointing it out to me! I've updated the code and now get the following behavior:

<div align='center'>

image

</div>

File: packages/task/src/browser/quick-open-task.ts [hunk 2]
```diff
@@ -312,6 +313,8 @@ export class TaskAttachQuickOpenItem extends QuickOpenItem {
 }
 export class TaskConfigureQuickOpenItem extends QuickOpenGroupItem {
 
+    protected taskDefinitionRegistry: TaskDefinitionRegistry;
```
Comment (RomanNikitenko) [inline packages/task/src/browser/quick-open-task.ts]: Where is it used?

File: packages/task/src/browser/quick-open-task.ts [hunk 3]
```diff
@@ -321,7 +324,9 @@ export class TaskConfigureQuickOpenItem extends QuickOpenGroupItem {
     }
 
     getLabel(): string {
-        return `${this.task._source}: ${this.task.label}`;
+        return this.task._source
```
Comment (RomanNikitenko) [inline packages/task/src/browser/quick-open-task.ts]: I guess for a configured task the name of the root folder will be used for displaying.
Please see [here](https://github.com/theia-ide/theia/blob/master/packages/task/src/common/task-protocol.ts#L49)
Is it expected behavior?

**Comment by RomanNikitenko** [comment]
@vince-fugnitto 
Could you test the case described [here](https://github.com/theia-ide/theia/pull/5472#discussion_r294477252)

**Comment by vince-fugnitto** [comment]
@RomanNikitenko sure thank you for pointing it out, I'll investigate it more closely.

File: packages/task/src/browser/quick-open-task.ts (hunk 3 above)
Comment (RomanNikitenko) [reply packages/task/src/browser/quick-open-task.ts]: I mean if you add some configuration to `tasks.json` file:
```
{
      "label": "Run my test task",
      "type": "shell",
      "command": "./task-long-running"
    }
```
For `Configure Tasks` the task is displayed :
configured_tasks_error

For `Run Task` the same task is displayed:
run_tasks

**Comment by vince-fugnitto** [comment]
I'm still not happy with the support present in a multi-root workspace.
The items should be grouped by `root`, have a `groupLabel` present, and have a divider.

_Currently_
<div align='center'>

Screen Shot 2019-08-07 at 11 48 33 AM


</div>

_VSCode_

<div align='center'>

Screen Shot 2019-08-07 at 10 22 25 AM


</div>

**Comment by vince-fugnitto** [comment]
@elaihau do you mind reviewing the PR for me when you get the chance?

File: packages/task/src/browser/quick-open-task.ts [hunk 4]
```diff
@@ -312,23 +319,39 @@ export class TaskAttachQuickOpenItem extends QuickOpenItem {
...
         protected readonly task: TaskConfiguration,
         protected readonly taskService: TaskService,
-        protected readonly labelProvider: LabelProvider
+        protected readonly labelProvider: LabelProvider,
+        protected readonly workspaceService: WorkspaceService,
+        protected readonly isMulti: boolean
     ) {
         super();
+        const stat = this.workspaceService.workspace;
+        this.isMulti = stat ? !stat.isDirectory : false;
```
Comment (akosyakov) [inline packages/task/src/browser/quick-open-task.ts]: you should not cache such things, but always compute them from `WorkspaceService` to make sure that you read up-to-date state, please review others places in the PR

Generally try to have a single source for the some state and read it from there, don't replicate it. Replication sometimes helpful, usually between processes, but introduces with keeping replica in sync and invalidating it properly.

Comment (akosyakov) [reply packages/task/src/browser/quick-open-task.ts]: `WorkspaceService` has `isMultiRootWorkspaceOpened`, just call it everywhere

**Comment by akosyakov** [comment]
@vince-fugnitto is not clear from the description how to test, please align with our template and assign more people for review

**Comment by vince-fugnitto** [comment]
> @vince-fugnitto is not clear from the description how to test, please align with our template and assign more people for review

Thanks, I update the PR description based on the new template.

**Comment by lmcbout** [comment]
Tested only the single root for now, I found the following when using the procedure "How to test" describe above:
If I delete the file: .theia/task.json --> the task remain in "Configure task" not in "detected task" unless I reload the Front-end
- When the file "task.json" is deleted, unable to recreate the file with Configure task unless I reload the Front-end

**Comment by vince-fugnitto** [comment]
> Tested only the single root for now, I found the following when using the procedure "How to test" describe above...

That is a behavior currently present in master and is not addressed in the scope of the PR.

**Comment by lmcbout** [comment]
Question:
When we already have some "Task" configured, should we show the list of "Task" as "Configured" and "Detected" when opening the dialog? Or may be have a separator between the already "configured" and the "detected" task ?

**Comment by vince-fugnitto** [comment]
> When we already have some "Task" configured, should we show the list of "Task" as "Configured" and "Detected" when opening the dialog? Or may be have a separator between the already "configured" and the "detected" task ?

I assume you mean when executing the `Configure Task` command which triggers the quick-open menu. VSCode does not display a separator between configured and detected tasks in this menu.

**Comment by lmcbout** [comment]
>I assume you mean when executing the Configure Task command which triggers the quick-open menu. VSCode does not display a separator between configured and detected tasks in this menu.

Yes in the Configure Task quick-open menu. I just tested with VSCode, you are right, but still I think a separator would make it better

**Comment by elaihau** [comment]
I tested the following things:
- both configured and detected tasks can be configured
- no duplications in `tasks.json` if a detected task is configured twice or more
- `recent tasks` can be configured
- works properly in a multi root workspace

**Comment by vince-fugnitto** [comment]
Thank you @RomanNikitenko @elaihau for the review.

**Pull request merged**